### PR TITLE
1-bit LAMB optimizer

### PR DESCRIPTION
**Pull request by conglongli**
1-bit LAMB: Communication Efficient Large-Scale Large-Batch Training with LAMB's Convergence Speed.

Author: @conglongli, @awan-10, @samyam, Hanlin Tang, Yuxiong He
Paper: https://arxiv.org/abs/2104.06069

### CODE REVIEW

File: docs/_pages/features.md [hunk 1]
```diff
@@ -172,15 +172,17 @@ Please see the [core API doc](https://deepspeed.readthedocs.io/) for more detail
 
 ## Training Optimizers
 
-### 1-bit Adam optimizer with up to 5x less communication
+### 1-bit Adam and 1-bit LAMB optimizers with up to 5x less communication
 
-DeepSpeed has an efficient implementation of a novel algorithm called 1-bit Adam.
-It offers the same convergence as Adam, incurs up to 5x less communication that enables
+DeepSpeed has two communication-efficient optimizers called 1-bit Adam and 1-bit LAMB.
+They offer the same convergence as Adam/LAMB, incurs up to 5x less communication that enables
```
Comment (awan-10) [inline docs/_pages/features.md:178]: minor typo: incurs--> incur

**Review by awan-10 — APPROVED** [review]

**Comment by awan-10** [comment]
Thank you @conglongli for this PR. It looks good to me. I just added a minor comment and approved all the other changes as is.

File: docs/_pages/features.md (hunk 1 above)
Comment (conglongli) [reply docs/_pages/features.md]: Fixed. Thanks Ammar.

File: docs/_pages/config-json.md [hunk 2]
```diff
@@ -88,6 +88,40 @@ The 1-bit Adam optimizer supports the following three params keys/values in addi
...
+
+```json
+"optimizer": {
+    "type": "OneBitLamb",
+    "params": {
+      "lr": 11e-3,
+      "weight_decay": 0.01,
+      "bias_correction": false,
+      "max_coeff": 0.3,
+      "min_coeff": 0.01,
```
Comment (jeffra) [inline docs/_pages/config-json.md:101]: should we add descriptions for max/min coeff in the table below?

**Review by jeffra — APPROVED** [review]
Can we use `self.deepspeed.enable_backward_allreduce` instead of creating `self.deepspeed.pipeline_enable_backward_allreduce`?

Otherwise looks good to me!

File: deepspeed/runtime/utils.py [hunk 3]
```diff
@@ -125,9 +130,17 @@ def has_overflow(self, params):
                                          op=torch.distributed.ReduceOp.MAX,
                                          group=torch.distributed.group.WORLD)
         elif self.mpu is not None:
+            if self.deepspeed.pipeline_enable_backward_allreduce is False:
```
Comment (jeffra) [inline deepspeed/runtime/utils.py:133]: This will crash if `self.deepspeed = None` right?

Comment (conglongli) [reply deepspeed/runtime/utils.py]: Yes you are right. I added checks about whether self.deepspeed exist, and whether it is PipelineEngine or not.

**Comment by conglongli** [comment]
> Can we use `self.deepspeed.enable_backward_allreduce` instead of creating `self.deepspeed.pipeline_enable_backward_allreduce`? /cc @ShadenSmith , @sdtblck
> 
> Otherwise looks good to me!

@jeffra @ShadenSmith We do need to add self.deepspeed.pipeline_enable_backward_allreduce in order to prevent PipelineEngine to do allreduce, because 1-bit Adam/LAMB it will handle the allreduce communication during the compression stage. On the other hand when using PipelineEngine we need to keep self.deepspeed.enable_backward_allreduce to always false since at 1-bit Adam/LAMB warmup stage we still need PipelineEngine to handle the allreduce. I did some further fix in 1-bit Adam/LAMB to make sure that when we use PipelineEngine, it will not touch self.deepspeed.enable_backward_allreduce, so that 1-bit Adam/LAMB shall not affect normal use of pipeline parallelism during warmup stage.

**Review by jeffra — APPROVED** [review]

File: docs/_pages/config-json.md (hunk 2 above)
Comment (conglongli) [reply docs/_pages/config-json.md:101]: Just added.

**Pull request merged**